flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -403,7 +403,10 @@ public void shutdown(JobStatus jobStatus) throws Exception {
 			// clear queued requests and in-flight checkpoints
 			abortPendingAndQueuedCheckpoints(reason);
 
-			completedCheckpointStore.shutdown(jobStatus);
+			// there might be a race condition with IO threads on completedCheckpointStore
+			synchronized (completedCheckpointStore) {
```

Comment:
If we are synchronizing all of the access to the `completedCheckpointStore` externally, wouldn't it be better to make `CompletedCheckpointStore` `@ThreadSafe` and synchronise it inside?

Reply:
Sounds reasonable.